dimod/core/bqm.py

```diff
@@ -1330,170 +1335,161 @@ def update(self, other):
+            return self
+        else:
+            return self._bqm.spin
 
-    def get_linear(self, v):
-        bqm = self._bqm
-        return 2 * bqm.get_linear(v) - 2 * sum(b for _, _, b in bqm.iter_quadratic(v))
+    def change_vartype(self, vartype, inplace=True):
+        if not inplace:
+            return self.copy().change_vartype(vartype, inplace=True)
 
-    def get_quadratic(self, u, v, default=None):
-        try:
-            return 4 * self._bqm.get_quadratic(u, v)
-        except ValueError as err:
-            if default is None:
-                raise err
-        return default
+        self.vartype = as_vartype(vartype)
 
-    def set_linear(self, v, bias):
-        bqm = self._bqm
+        return self
 
-        # need the difference
-        delta = bias - self.get_linear(v)
+    def copy(self):
+        return self._bqm.change_vartype(self.vartype, inplace=False)
 
-        bqm.set_linear(v, bqm.get_linear(v) + delta / 2)
-        bqm.offset += delta / 2
+    def degree(self, *args, **kwargs):
+        return self._bqm.degree(*args, **kwargs)
 
-    def set_quadratic(self, u, v, bias):
+    def get_linear(self, v):
         bqm = self._bqm
+        if self.vartype is bqm.vartype:
+            return bqm.get_linear(v)
+        elif self.vartype is Vartype.SPIN and bqm.vartype is Vartype.BINARY:
+            return bqm.get_linear(v) / 2 + bqm.adj[v].sum() / 4
+        elif self.vartype is Vartype.BINARY and bqm.vartype is Vartype.SPIN:
+            return 2 * bqm.get_linear(v) - 2 * bqm.adj[v].sum()
+        else:
+            raise RuntimeError("unknown vartype combination")
 
-        # need the difference
-        delta = bias - self.get_quadratic(u, v, default=0)
-
-        # if it doesn't exist and BQM is not shapeable, this will fail
-        bqm.set_quadratic(u, v, bias / 4)  # this one is easy
-
-        # the other values get the delta
-        bqm.set_linear(u, bqm.get_linear(u) + delta / 4)
-        bqm.set_linear(v, bqm.get_linear(v) + delta / 4)
+    def get_quadratic(self, u, v, default=None):
+        bqm = self._bqm
+        if self.vartype is bqm.vartype:
+            return bqm.get_quadratic(u, v, default=default)
+        elif self.vartype is Vartype.SPIN and bqm.vartype is Vartype.BINARY:
+            return bqm.get_quadratic(u, v) / 4
+        elif self.vartype is Vartype.BINARY and bqm.vartype is Vartype.SPIN:
+            return 4 * self._bqm.get_quadratic(u, v)
+        else:
+            raise RuntimeError("unknown vartype combination")
 
-        bqm.offset += delta / 4
+    def iter_linear(self):
+        for v, _ in self._bqm.iter_linear():
+            yield v, self.get_linear(v)
 
+    def iter_quadratic(self, variables=None):
+        for u, v, _ in self._bqm.iter_quadratic(variables):
+            yield u, v, self.get_quadratic(u, v)
 
-class SpinView(VartypeView):
-    @property
-    def binary(self):
-        return self._bqm.binary
+    def relabel_variables(self, *args, **kwargs):
```

Comment:
This seems wrong because it can modify the bqm beneath the view.

```python
def relabel_variables(self, mapping, inplace=True):
    relabeled = self._bqm.relabel_variables(mapping, inplace=False)
    if inplace:
        self._bqm = relabeled
    return relabeled
```